lib/ohai/mixin/azure_metadata.rb

```diff
@@ -19,30 +19,87 @@
+    # work and executes the metadata retrieval using that version.
+    #
+    # If no compatible version is found, an empty hash is returned.
+    #
     module AzureMetadata
 
       AZURE_METADATA_ADDR ||= "169.254.169.254".freeze
-      AZURE_METADATA_URL ||= "/metadata/instance?api-version=2017-08-01".freeze
 
-      # fetch the meta content with a timeout and the required header
+      # it's important that the newer versions are at the end of this array so we can skip sorting it
+      AZURE_SUPPORTED_VERSIONS ||= %w{ 2017-04-02 2017-08-01 2017-12-01 2018-02-01 2018-04-02
+                                   2018-10-01 2019-02-01 2019-03-11 2019-04-30 2019-06-01
+                                   2019-06-04 2019-08-01 2019-08-15 2019-11-01 }.freeze
+
+      def best_api_version
+        @api_version ||= begin
+          logger.trace("Mixin AzureMetadata: Fetching http://#{AZURE_METADATA_ADDR}/metadata/instance to determine the latest supported metadata release")
+          response = http_get("/metadata/instance")
+          if response.code == "404"
+            logger.trace("Mixin AzureMetadata: Received HTTP 404 from metadata server while determining API version, assuming #{AZURE_SUPPORTED_VERSIONS.last}")
+            return AZURE_SUPPORTED_VERSIONS.last
+          elsif response.code != "400" # 400 is actually what we want
+            raise "Mixin AzureMetadata: Unable to determine Azure metadata version (returned #{response.code} response)"
+          end
+
+          # azure returns a list of the 3 latest versions it supports
+          versions = parse_json(response.body)["newest-versions"]
+          versions.sort!
+
+          until versions.empty? || AZURE_SUPPORTED_VERSIONS.include?(versions.last)
+            pv = versions.pop
+            logger.trace("Mixin AzureMetadata: Azure metadata version #{pv} is not present in the versions provided by the Azure Instance Metadata service")
+          end
+
+          if versions.empty?
+            logger.debug "Mixin AzureMetadata: The short list of supported versions provided by Azure Instance Metadata service doesn't match any known versions to Ohai. Using the latest supported release known to Ohai instead: #{AZURE_SUPPORTED_VERSIONS.last}"
+            return AZURE_SUPPORTED_VERSIONS.last
+          end
+
+          logger.trace("Mixin AzureMetadata: Latest supported Azure metadata version: #{versions.last}")
+          versions.last
+        end
+      end
+
+      # fetch the meta content with a timeout and the required header and a read timeout of 6s
+      #
+      # @param [String] the relative uri to fetch from the Azure Metadata Service URL
+      #
+      # @return [Net::HTTP]
       def http_get(uri)
+        full_uri = URI.join("http://#{AZURE_METADATA_ADDR}", uri)
         conn = Net::HTTP.start(AZURE_METADATA_ADDR)
         conn.read_timeout = 6
-        conn.get(uri, { "Metadata" => "true" })
+        conn.get(full_uri, { "Metadata" => "true" })
```

Comment:
I would have expected this to work with just the path, does that not work?

Reply:
Azure requires you to set a header or it fails

Reply:
I meant just the path rather than constructing the URL out of the host + path.

Reply:
oh yep I fixed that